### PR TITLE
Solicitor enters arrival time and selects today/tomorrow for date

### PR DESCRIPTION
**Pull request by robmckinnon**
* JS `DateChooser` class can be reused and extended for CSO interview date selection UI.
* Partially implements story: https://www.pivotaltracker.com/story/show/90514388
* `DateChooser` allows Today/Tomorrow toggling - entering a date will be done in another story.
* Non-JS version defaults to the existing three input date entry.
* Add Jasmine ad for JS unit testing.
* Add Jasmine and Rubocop checks to Guardfile for local testing.
* Add Jasmine CI task to default rake task - which is run by CircleCI.
* Arrival time setting actions moved into new `SolicitorArrivalTimesController`.
* Removed `terminal-notifier-guard`. I appear to be only one using Guard at present, and didn't need OS notifications.
* Implemented JS code in coffeescript, as agreed with team on Friday.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -44,6 +43,13 @@ group :development, :test do
   gem "quiet_assets", "~> 1.1"
   gem "rspec-rails", "~> 3.2.0"
   gem "rubocop", "~> 0.30"
+  gem "phantomjs", "~> 1.9.7"
+  gem "jasmine-core", "~> 2.0.0"
+  gem "jasmine", "~> 2.0.2"
+  gem "jasmine-rails", "~> 0.10.2"
+  gem "jasmine-jquery-rails", "~> 2.0.2"
+  gem "guard-jasmine", git: "https://github.com/guard/guard-jasmine", branch: "jasmine-2"
```
Comment (chriscarter90) [inline Gemfile]: Was there not an on-going discussion about whether personal gem preferences should remain in the Gemfile? If it's not on-going, what was the decision?

File: app/controllers/concerns/defence_request_authorization.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  def set_policy_with_context
+    @policy ||= policy(PolicyContext.new(defence_request, current_user))
+  end
+
+  def defence_request
+    @defence_request ||= DefenceRequest.new
+  end
+
+  def new_defence_request_form
+    @defence_request_form ||= DefenceRequestForm.new @defence_request
```
Comment (chriscarter90) [inline app/controllers/concerns/defence_request_authorization.rb]: Seems to be a mix of style here - calling methods with no brackets

File: app/helpers/defence_requests_helper.rb [hunk 3]
```diff
@@ -4,4 +4,18 @@ def solicitor_id solicitor_hash
     "solicitor-#{solicitor_hash['id']}"
   end
 
+  def data_chooser_setup date_to_edit
```
Comment (chriscarter90) [inline app/helpers/defence_requests_helper.rb]: ()

File: spec/features/solicitor/defence_request_management_spec.rb [hunk 4]
```diff
@@ -2,9 +2,20 @@
 
 RSpec.feature "Solicitors managing defence requests" do
   context "with cases they are assigned to" do
-    specify "can see the show page of the request" do
+
+    def enter_time hour:, min:, day: nil, month: nil, year: nil
```
Comment (chriscarter90) [inline spec/features/solicitor/defence_request_management_spec.rb]: ()

File: Gemfile (hunk 1 above)
Comment (robmckinnon) [inline Gemfile]: I asked team last week if Jasmine was ok for unit testing javascript. If including guard is causing issues please let me know.

File: app/controllers/concerns/defence_request_authorization.rb (hunk 2 above)
Comment (robmckinnon) [inline app/controllers/concerns/defence_request_authorization.rb]: We could add this check to rubocop if you want it enforced.

File: Gemfile [hunk 5]
```diff
@@ -44,6 +43,13 @@ group :development, :test do
   gem "quiet_assets", "~> 1.1"
   gem "rspec-rails", "~> 3.2.0"
   gem "rubocop", "~> 0.30"
+  gem "phantomjs", "~> 1.9.7"
+  gem "jasmine-core", "~> 2.0.0"
```
Comment (FriedSock) [inline Gemfile]: What is `jasmine-core`, and why do we need it as well as `jasmine`?

File: app/helpers/defence_requests_helper.rb (hunk 3 above)
Comment (robmckinnon) [inline app/helpers/defence_requests_helper.rb]: We could also add this check to rubocop if you want it enforced.

File: Gemfile (hunk 5 above)
Comment (robmckinnon) [inline Gemfile]: It took a while to get a stable combination of gem versions to run `jasmine` with `phantomjs`.  `jasmine-core` is a dependency required by `jasmine`:
```ruby
    jasmine (2.0.3)
      jasmine-core (~> 2.0.0)
```
I'll remove the explicit require, and commit if everything runs ok.

**Comment by robmckinnon** [comment]
There has been discussion among developers relating to how the non-javascript date entry UI should look and behave, and how date entry should look and behave when JS is enabled.

Happy to do further implementation work once we have clarified the design with Kate.

**Comment by robmckinnon** [comment]
Discussed with @ansonK. Agreed for now we will show three field date entry when JS is enabled.

File: app/controllers/defence_requests_controller.rb [hunk 6]
```diff
@@ -1,8 +1,10 @@
 class DefenceRequestsController < BaseController
 
+  include DefenceRequestAuthorization
+
```
Comment (dp6ai) [inline app/controllers/defence_requests_controller.rb]: as discussed with @FriedSock and @robmckinnon i think this is a good refactor if we can make it generic enough to use across all the controllers

**Comment by robmckinnon** [comment]
As discussed with @FriedSock and @dp6ai refactoring to make more use of defence request authorization module mixin across controllers.

**Comment by dp6ai** [comment]
LGTM

**Comment by robmckinnon** [comment]
@dp6ai thanks for the feedback. Made changes as suggested. Merging.

File: app/helpers/defence_requests_helper.rb (hunk 3 above)
Comment (chriscarter90) [inline app/helpers/defence_requests_helper.rb]: This has now been enforced in Rubocop.

**Pull request merged**